srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1241 @@
+				let value = original_free_balance.checked_add(&amount)
+					.ok_or_else(|| "free balance got overflow after minting.")?;
+
+				<TotalIssuance<T>>::insert(asset_id, new_total_issuance);
+				Self::set_free_balance(&asset_id, &to, value);
+
+				Self::deposit_event(RawEvent::Minted(asset_id, to, amount));
+
+				Ok(())
+			} else {
+				Err("The origin does not have permission to mint an asset.")
+			}
+		}
+
+		/// Burns an asset, decreases its total issuance.
+		///
+		/// The `origin` must have `burn` permissions.
+		fn burn(origin, #[compact] asset_id: T::AssetId, to: T::AccountId, amount: T::Balance) -> Result {
+			let origin = ensure_signed(origin)?;
+
+			if Self::check_permission(&asset_id, &origin, &PermissionType::Burn) {
+				let original_free_balance = Self::free_balance(&asset_id, &to);
+
+				let current_total_issuance = <TotalIssuance<T>>::get(asset_id);
+				let new_total_issuance = current_total_issuance.checked_sub(&amount)
+					.ok_or_else(|| "total_issuance got underflow after burning")?;
+				let value = original_free_balance.checked_sub(&amount)
+					.ok_or_else(|| "free_balance got underflow after burning")?;
+
+				<TotalIssuance<T>>::insert(asset_id, new_total_issuance);
+
+				Self::set_free_balance(&asset_id, &to, value);
+
+				Self::deposit_event(RawEvent::Burned(asset_id, to, amount));
+
+				Ok(())
+			} else {
+				Err("The origin does not have permission to burn an asset.")
+			}
+		}
+
+		/// Can be used to create reserved tokens.
+		/// Requires Root call.
+		fn create_reserved(asset_id: T::AssetId, options: AssetOptions<T::Balance, T::AccountId>) -> Result {
+			Self::create_asset(Some(asset_id), None, options)
+		}
+	}
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct BalanceLock<Balance, BlockNumber> {
+	pub id: LockIdentifier,
+	pub amount: Balance,
+	pub until: BlockNumber,
+	pub reasons: WithdrawReasons,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as GenericAsset {
+		/// Total issuance of a given asset.
+		pub TotalIssuance get(total_issuance) build(|config: &GenesisConfig<T>| {
+			let issuance = config.initial_balance * (config.endowed_accounts.len() as u32).into();
+			config.assets.iter().map(|id| (id.clone(), issuance)).collect::<Vec<_>>()
+		}): map T::AssetId => T::Balance;
+
+		/// The free balance of a given asset under an account.
+		pub FreeBalance: double_map T::AssetId, twox_128(T::AccountId) => T::Balance;
+
+		/// The reserved balance of a given asset under an account.
+		pub ReservedBalance: double_map T::AssetId, twox_128(T::AccountId) => T::Balance;
+
+		/// Next available ID for user-created asset.
+		pub NextAssetId get(next_asset_id) config(): T::AssetId;
+
+		/// Permission options for a given asset.
+		pub Permissions get(get_permission): map T::AssetId => PermissionVersions<T::AccountId>;
+
+		pub CreateAssetStakes get(create_asset_stake) config(): T::Balance;
```

Comment:
missing docs.

Reply:
I will just remove this because it is too specific to our network. Will later introduce some generic way for runtime to customize the criterial to create a new asset.